### ISSUE TITLE
Wayland support?

### ISSUE THREAD

**Issue by micush** [openrecall/openrecall#10]
Hi,

This doesn't seem to work on Wayland. MSS appears to only support X11. Any plans to implement Wayland support?

-micush

**Comment by koenvaneijk** [comment]
Thanks @micush for your suggestion. Given the number of upvotes this is getting, we're adding this one to the roadmap. I will update this issue to reflect progress.

**Comment by koenvaneijk** [comment]
## Feature: Wayland Support

### Description
Currently, OpenRecall leverages the MSS library, which supports screenshot capturing on X11. However, with the increasing adoption of Wayland, it's crucial to ensure OpenRecall is compatible with this display server protocol. Implementing Wayland support will enable users on Wayland-based systems, such as recent versions of GNOME and KDE, to use OpenRecall without any limitations.

### Implementation Steps

1. **Research Wayland Protocols**:
   - Investigate the Wayland protocols and libraries that support screenshot capturing.
   - Libraries such as `wlroots`, `wf-recorder`, and `grim` can be evaluated for their capabilities.

2. **Select a Suitable Library**:
   - Choose a library that is well-maintained and supports the required functionality for capturing screenshots on Wayland.

3. **Abstract Screenshot Capture**:
   - Refactor the existing codebase to abstract the screenshot capturing logic.
   - Implement a detection mechanism to check if the system is running X11 or Wayland.

4. **Implement Wayland Capture**:
   - Develop a module that uses the selected Wayland library to capture screenshots.
   - Ensure that this module integrates seamlessly with the existing application logic.

5. **Testing**:
   - Thoroughly test the implementation on various Wayland compositors (e.g., GNOME Shell, KDE Plasma).
   - Ensure the feature works reliably without impacting performance.

6. **Documentation**:
   - Update the project documentation to include details about Wayland support.
   - Provide instructions for users on Wayland-based systems on how to configure and use OpenRecall.

### Benefits
- **Increased Compatibility**: Users of modern Linux distributions that default to Wayland will be able to use OpenRecall without needing to switch to X11.
- **Future-proofing**: Supporting Wayland ensures that OpenRecall remains relevant as more distributions and environments move away from X11.
- **Enhanced User Experience**: Users will enjoy a seamless experience regardless of their underlying display server protocol.

**Comment by lavafroth** [comment]
For the sake of continuity, in #27, I recommended pyscreenshot for taking screenshots on wayland. For now, [here](https://github.com/lavafroth/openrecall)'s my fork with wayland support.

**Comment by josh-at-straker** [comment]
Just to add to this, for Gnome users, it would be great to have something other than `gnome-screenshot` as the backend on wayland.

**Comment by khyerdev** [comment]
for the wayland/x11 detection, I believe that the display compositor / window manager sets the environment variable XDG_SESSION_TYPE to hold either "x11" or "wayland", which is available for applications running under the compositor.

**Comment by khyerdev** [comment]
i would also suggest also testing on a wayland dynamic window manager like Hyprland, as it is gaining more users (and is currently the most popular tiling window manager), and not just on Gnome and KDE. I say this because the maintainer of Hyprland, Vaxry, is rewriting ALL of wlroots in C++, along with the implementations of it and other FDO wayland stuff.
I can also recommend grim for the screenshotting, as I use it for taking screenshots. it is really fast.

**Comment by nonetrix** [comment]
> ## Feature: Wayland Support
> ### Description
> Currently, OpenRecall leverages the MSS library, which supports screenshot capturing on X11. However, with the increasing adoption of Wayland, it's crucial to ensure OpenRecall is compatible with this display server protocol. Implementing Wayland support will enable users on Wayland-based systems, such as recent versions of GNOME and KDE, to use OpenRecall without any limitations.
> 
> ### Implementation Steps
> 1. **Research Wayland Protocols**:
>    
>    * Investigate the Wayland protocols and libraries that support screenshot capturing.
>    * Libraries such as `wlroots`, `wf-recorder`, and `grim` can be evaluated for their capabilities.
> 2. **Select a Suitable Library**:
>    
>    * Choose a library that is well-maintained and supports the required functionality for capturing screenshots on Wayland.
> 3. **Abstract Screenshot Capture**:
>    
>    * Refactor the existing codebase to abstract the screenshot capturing logic.
>    * Implement a detection mechanism to check if the system is running X11 or Wayland.
> 4. **Implement Wayland Capture**:
>    
>    * Develop a module that uses the selected Wayland library to capture screenshots.
>    * Ensure that this module integrates seamlessly with the existing application logic.
> 5. **Testing**:
>    
>    * Thoroughly test the implementation on various Wayland compositors (e.g., GNOME Shell, KDE Plasma).
>    * Ensure the feature works reliably without impacting performance.
> 6. **Documentation**:
>    
>    * Update the project documentation to include details about Wayland support.
>    * Provide instructions for users on Wayland-based systems on how to configure and use OpenRecall.
> 
> ### Benefits
> * **Increased Compatibility**: Users of modern Linux distributions that default to Wayland will be able to use OpenRecall without needing to switch to X11.
> * **Future-proofing**: Supporting Wayland ensures that OpenRecall remains relevant as more distributions and environments move away from X11.
> * **Enhanced User Experience**: Users will enjoy a seamless experience regardless of their underlying display server protocol.

This sounds suspiciously AI generated 🤔

**Comment by yacc143** [comment]
> i would also suggest also testing on a wayland dynamic window manager like Hyprland, as it is gaining more users (and is currently the most popular tiling window manager), and not just on Gnome and KDE. I say this because the maintainer of Hyprland, Vaxry, is rewriting ALL of wlroots in C++, along with the implementations of it and other FDO wayland stuff. I can also recommend grim for the screenshotting, as I use it for taking screenshots. it is really fast.

 `(base) andreas@idefix:~$ grim
compositor doesn't support wlr-screencopy-unstable-v1
`
Not wanting to complain much, but grim does not support KDE on wayland. Not very acceptable.

pyscreenshot will use grim internally if necessary and installed:

`(base) andreas@idefix:~$ pipx run --pip-args "Pillow pydbus" --spec pyscreenshot python -m pyscreenshot.check.speedtest
⚠  python is already on your PATH and installed at /usr/bin/python. Downloading and running anyway.

n=10
------------------------------------------------------
default                 6    sec        (  596 ms per call)
pil                 
mss                 
scrot               
grim                
maim                
imagemagick         
pyqt5               
pyside2             
wx                  
pygdk3              
mac_screencapture   
mac_quartz          
gnome_dbus          
gnome-screenshot    `

**Comment by yacc143** [comment]
> For the sake of continuity, in #27, I recommended pyscreenshot for taking screenshots on wayland. For now, [here](https://github.com/lavafroth/openrecall)'s my fork with wayland support.

Having looked at your fork, it looks nice, but it has one primary problem, which stems from one feature that is missing from pyscreenshot:

- pyscreenshot defaults to taking the whole screen as a screenshot.
- it can also take parts of the whole screen by specifying a bounding box.
- it has no knowledge of monitors/displays.
- so it's not possible without additional tools to take a screenshot of the primary monitory only → so the primary monitory only option is not directly implementable.

**Comment by yacc143** [comment]
>     * so it's not possible without additional tools to take a screenshot of the primary monitory only → so the primary monitory only option is not directly implementable.

On Linux, even on wayland, running xrandr spits out the screens:

`(base) andreas@idefix:~$ xrandr | grep connected
#eDP-1 connected primary 1536x864+2560+576 (normal left inverted right x axis y axis) 309mm x 173mm
DP-1 connected 2560x1440+0+0 (normal left inverted right x axis y axis) 698mm x 393mm
`

**Comment by dotsupershow** [comment]
Yes! I second the wayland support and using something other than Gnome screenshots.

**Comment by aamaanaa** [comment]
Yes! i third it for support on wayland as fedora 43 dropped x11
